### PR TITLE
New function: `hasPlayerSpawned`

### PR DESCRIPTION
**Pull request by TracerDS**
Returns `true` when player has been spawned by using `spawnPlayer`. `false` otherwise

### CODE REVIEW

File: Client/mods/deathmatch/logic/CClientPed.cpp
Comment (tederis) [inline Client/mods/deathmatch/logic/CClientPed.cpp]: I would avoid the expressions like this. Looks clumsy and violates the design principles. Have you thought about moving the entire method CClientPed::Spawn under the CClientPlayer class? Or

Comment (TracerDS) [reply Client/mods/deathmatch/logic/CClientPed.cpp]: Cant move it to the player. Players inherit functions from peds and `Spawn` manage the ped's being alive. If I move it to CClientPlayer then I will break CPed

Comment (tederis) [reply Client/mods/deathmatch/logic/CClientPed.cpp]: The only place from which this method is invoked is Packet_PlayerSpawn which on its own is dedicated for players only(not peds). It cannot break non-player peds.

Comment (TracerDS) [reply Client/mods/deathmatch/logic/CClientPed.cpp]: > The only place from which this method is invoked is Packet_PlayerSpawn which on its own is dedicated for players only(not peds). It cannot break non-player peds.

Maybe instead of moving the entire method, I'll move only `SetPlayerSpawned` there
```cpp
pPlayer->Spawn(vecPosition, fRotation, usPlayerModelID, ucInterior);
pPlayer->SetPlayerSpawned(true);
```

Comment (tederis) [reply Client/mods/deathmatch/logic/CClientPed.cpp]: That's better. But breaks the object's state coherency. Because someone in the future can invoke Spawn but forget to call SetPlayerSpawned and boom, object's state is invalid now.

Comment (TracerDS) [reply Client/mods/deathmatch/logic/CClientPed.cpp]: Okay, I'll move it to `CClientPlayer`

Comment (tederis) [reply Client/mods/deathmatch/logic/CClientPed.cpp]: I would suggest you to leave it as is. But after the merge create a new PR with the corresponding refactor.

Comment (TracerDS) [reply Client/mods/deathmatch/logic/CClientPed.cpp]: If it wont cause any issues then its better to do it now instead of waiting for it 'til merge

File: Server/mods/deathmatch/logic/CPlayer.h
Comment (tederis) [inline Server/mods/deathmatch/logic/CPlayer.h]: There already exists field CPed::m_bSpawned on server side. I know it behaves a bit different, but maybe there could be a way to avoid that collision. Because two fields with almost the same name may be confusing.

File: Server/mods/deathmatch/logic/CMapManager.cpp
Comment (tederis) [inline Server/mods/deathmatch/logic/CMapManager.cpp]: I see that SetPlayerSpawned is called only ones. Should't there be a place where you call `SetPlayerSpawned(false)`?

Comment (TracerDS) [reply Server/mods/deathmatch/logic/CMapManager.cpp]: > I see that SetPlayerSpawned is called only ones. Should't there be a place where you call `SetPlayerSpawned(false)`?

You can't despawn a player

File: Server/mods/deathmatch/logic/CPlayer.h
Comment (TracerDS) [reply Server/mods/deathmatch/logic/CPlayer.h]: > There already exists field CPed::m_bSpawned on server side. I know it behaves a bit different, but maybe there could be a way to avoid that collision. Because two fields with almost the same name may be confusing.

`m_bHasPlayerSpawned` will change to `false` when player dies. If I were to make these 2 variables into 1, I'd need to make a complete overhaul of the spawn system

File: Server/mods/deathmatch/logic/CMapManager.cpp
Comment (tederis) [reply Server/mods/deathmatch/logic/CMapManager.cpp]: > > I see that SetPlayerSpawned is called only ones. Should't there be a place where you call `SetPlayerSpawned(false)`?
> 
> You can't despawn a player

What will happened when player is died then?

Comment (TracerDS) [reply Server/mods/deathmatch/logic/CMapManager.cpp]: > What will happened when player is died then?

When player dies, `m_bHasPlayerSpawned` will change to `false` (indicating that the player has been despawned) but the player itself wont despawn, he will retain every property of a spawned player.
Its kinda confusing tbf. Dead players should be dead, not despawned.

Comment (TracerDS) [reply Server/mods/deathmatch/logic/CMapManager.cpp]: Actually, maybe I should just update `SetSpawned` to only work for spawning player and retain `SetDead` for killing/reviving players/peds? That would make more sense, no?

Comment (tederis) [reply Server/mods/deathmatch/logic/CMapManager.cpp]: > > What will happened when player is died then?
> 
> When player dies, `m_bHasPlayerSpawned` will change to `false` (indicating that the player has been despawned) but the player itself wont despawn, he will retain every property of a spawned player. Its kinda confusing tbf. Dead players should be dead, not despawned.

But how exactly the value of m_bHasPlayerSpawned can become false? There's no place where it's set to false in your code.

Comment (TracerDS) [reply Server/mods/deathmatch/logic/CMapManager.cpp]: > > > What will happened when player is died then?
> > 
> > 
> > When player dies, `m_bHasPlayerSpawned` will change to `false` (indicating that the player has been despawned) but the player itself wont despawn, he will retain every property of a spawned player. Its kinda confusing tbf. Dead players should be dead, not despawned.
> 
> But how exactly the value of m_bHasPlayerSpawned can become false? There's no place where it's set to false in your code.

Oops, sorry. I meant `m_bSpawned` (the original variable that changes when player dies)

`m_bHasPlayerSpawned` should only be set to false when player joins and leaves.
Since there is no direct way to despawn a player, it will never be set to false once a player has been spawned.

Comment (tederis) [reply Server/mods/deathmatch/logic/CMapManager.cpp]: > Actually, maybe I should just update `SetSpawned` to only work for spawning player and retain `SetDead` for killing/reviving players/peds? That would make more sense, no?

Yeah. I think `m_bSpawned` could be renamed to m_bDead (with the corresponding refactor, of course). When player connects m_bDead is false, when he's dead m_bDead is true (obviously), and when he's spawned after it m_bDead is false again. In this case we can avoid the collision and there will no any confusion.

File: Server/mods/deathmatch/logic/CGame.cpp
Comment (tederis) [inline Server/mods/deathmatch/logic/CGame.cpp:3809]: I see you has taken care of two call sites where IsSpawned was used. But there are more of them: 
spawned

Comment (TracerDS) [reply Server/mods/deathmatch/logic/CGame.cpp:3809]: I didnt know if I should replace them since these lines operate on alive peds

Comment (tederis) [reply Server/mods/deathmatch/logic/CGame.cpp:3809]: That's important because you have changed the meaning of `m_bSpawned`.

Comment (tederis) [reply Server/mods/deathmatch/logic/CGame.cpp:3809]: It's become complicated. What if you revert the last commit and just rename m_bHasPlayerSpawned with something like m_bAlive? Simple and no collisions.

Comment (TracerDS) [reply Server/mods/deathmatch/logic/CGame.cpp:3809]: > That's important because you have changed the meaning of `m_bSpawned`.

I can change `IsSpawned` to `IsDead`. Based on the usage of the `SetSpawned` before the PR, `SetSpawned` is the same as `SetIsDead(not IsDead())`

File: Server/mods/deathmatch/logic/CGame.cpp
Comment (tederis) [inline Server/mods/deathmatch/logic/CGame.cpp:4372]: You have replaced IsSpawned with !IsDead(). But at this line you can see that they are not interchangeable. Because m_bIsDead is only set when player is wasted. But m_bSpawned is true when player's connection is completed, not only when he's spawned via spawnPlayer.

Comment (TracerDS) [reply Server/mods/deathmatch/logic/CGame.cpp:4372]: that line of code itself doesnt make sense. When player joins a server (lets say without any resources) then he only joined but he is not spawned. Only `spawnPlayer` will spawn a player. Currently, mta sets `m_bSpawned` to the opposite of `m_bIsDead` on every dying related event or function.

Comment (tederis) [reply Server/mods/deathmatch/logic/CGame.cpp:4372]: Okay. But it should be noted that these changes will break the backward compatibility. Lets take a look at `setElementHealth` for example. Currently `setElementHealth` is valid even if it's called before `spawnPlayer`. After your changes it will not be true anymore.

Comment (tederis) [reply Server/mods/deathmatch/logic/CGame.cpp:4372]: On the other hand, who the hell would want to use setElementHealth if player is not even spawned, right?

Comment (TracerDS) [reply Server/mods/deathmatch/logic/CGame.cpp:4372]: > On the other hand, who the hell would want to use setElementHealth if player is not even spawned, right?

It also *might potentially* fix #472 but I really doubt it.

Backwards compatibility wouldn't really be a problem here. A quick refactor of `setElementHealth` to return `false` when the player is not spawned

File: Server/mods/deathmatch/logic/luadefs/CLuaPlayerDefs.h
Comment (FileEX) [inline Server/mods/deathmatch/logic/luadefs/CLuaPlayerDefs.h]: player can be const?

File: Server/mods/deathmatch/logic/luadefs/CLuaPlayerDefs.cpp
Comment (FileEX) [inline Server/mods/deathmatch/logic/luadefs/CLuaPlayerDefs.cpp]: player can be const?

File: Client/mods/deathmatch/logic/CClientPlayer.cpp
Comment (FileEX) [inline Client/mods/deathmatch/logic/CClientPlayer.cpp]: nullptr?

File: Client/mods/deathmatch/logic/luadefs/CLuaPlayerDefs.cpp
Comment (FileEX) [inline Client/mods/deathmatch/logic/luadefs/CLuaPlayerDefs.cpp]: noexcept?

File: Server/mods/deathmatch/logic/luadefs/CLuaPlayerDefs.cpp
Comment (FileEX) [reply Server/mods/deathmatch/logic/luadefs/CLuaPlayerDefs.cpp]: noexcept?

**Review by FileEX — APPROVED** [review]

File: Server/mods/deathmatch/logic/luadefs/CLuaPlayerDefs.h
Comment (TracerDS) [reply Server/mods/deathmatch/logic/luadefs/CLuaPlayerDefs.h]: ~~When I marked any pointer `const`, ArgumentParser rejected these functions and threw a compile time error~~

Right, thanks

File: Server/mods/deathmatch/logic/CGame.cpp
Comment (Dutchman101) [reply Server/mods/deathmatch/logic/CGame.cpp:4372]: I've unresolved this review because as it stands, it looks like the backwards compatibility concern hasn't been tackled.
`A quick refactor of setElementHealth to return false when the player is not spawned` (@TracerDS)
Then please open a PR to do that, which will be a blocker for this to resolve and merge.

@tederis Do you have anything to add?

Comment (Dutchman101) [reply Server/mods/deathmatch/logic/CGame.cpp:4372]: @TracerDS I see you're updating the PR, but not acknownledging the above review.
Can you clarify why, because otherwise merge isn't gonna happen.

Comment (TracerDS) [reply Server/mods/deathmatch/logic/CGame.cpp:4372]: > @TracerDS I see you're updating the PR, but not acknownledging the above review. Can you clarify why, because otherwise merge isn't gonna happen.

`setElementHealth` currently checks if player was spawned.
The value of `IsSpawned()` is the same as `!IsDead()`.
They were interchangeable. `m_bIsDead` is set to `false` in `CPed`'s constructor.
`m_bIsSpawned` is set to `true` **only** here.
So in current mta's code its always set to `true`.

In this PR you cannot set player's health when he is not connected to the server yet.
If it stayed the same then it might cause some issues.
Checking against `IsDead` instead fixes these issues and returns false when ped is dead/not spawned.
However, that would also mean that using `setElementHealth` on a dead ped will fail (return false). If that is a concern, I could remove the whole `if` statement here since its always set to `true` in previous code

**Review by Fernando-A-Rocha — APPROVED** [review]
Good job

**Comment by TracerDS** [comment]
Im not interested in advancing this PR further anymore.
If someone wants to refresh it, feel free to do so.

**Pull request closed without merging**
